### DISCUSSION TITLE
Make Solid.JS a fully featured Production Framework

### DISCUSSION THREAD

**Discussion by CordialGit** [solidjs/solid#1772] [Ideas]
Normally, JavaScript is powerful. It can do so much on the frontend. And it is a sufficient tool especially when paired with PHP, Node.Js or Deno.

Now, I'm offended that creators of JavaScript frameworks are now making Frontend Development too difficult.

**THE PROBLEM**
If I've taken the pain to learn HTML, CSS and JavaScript(_which on its own, is powerful on the frontend_).  Then I'm going forward to learn a Framework like Solid.JS, or  Vue or React(**for example**).

Why can't any of these, just be enough?

But No! -- I still keep being forced to learn much more!

In the case of Vue, I still need to learn Nuxt for Server Side Rendering, Pinia for State Management, Vuetify for Components, and Vitest for testing?

In the case of React, I still need to learn Next for Server Side Rendering, Flow or Redux for State Management, maybe Material UI for Components, maybe Styled Components for CSS-in-JS styling, and Jest for testing.

All of these over-learning is making frontend Development too difficult. They make frontend Development to become too complex and too difficult.

With so much frontend power in JavaScript already, can't these JavaScript Frameworks just come with those extra additional features in-built?

My wish now, is that a new framework will come along the way, and it will fix all these problems of Complexity. 

It's frustrating learning too much and learning some new tools because of some ordinary small additional features--- small features that can simply be built into the Framework's Core.

**THE ALTERNATIVE**
I make websites with WordPress, and all the headaches that coding from scratch presents, are all taken away--while using WordPress.

With WordPress, a Database is already there. A Content creation Interface is already there. Admin Dashboard for managing Themes(Looks) and Plugins(Functionality) is already there. Everything you need is all there, and it's simply Drag-and-Drop to build your website.

The WordPress process is so easy, and it keeps evolving to make life and website development so easy.

**MY REQUEST TO SOLID.JS**
Make Solid.JS to become a fully featured **Batteries-included** framework like WordPress or Python.

This Fully Featured Batteries-Included Framework concept for Solid.JS would mean that, after the pain of learning JavaScript, and we then add Solid.JS on top of that JavaScript, we should need nothing more, as far as frontend is concerned.

So that all the things that Nuxt brings to Vue, or all the things that Next brings to React, should all be taken care of, by Solid.JS once and for all. 

And we should not need again, another Framework on top of Solid.JS

**IN PASSING**
With a **Batteries-Included** Fully Featured approach for Solid.Js, it will move from being a "_simple and performant reactivity for building user interfaces_" to being called "_the JavaScript Framework for Production Ready Apps_".

In essence, I am asking that Solid.JS  should make our life easier by being a fully featured Framework with a Batteries-Included approach--- where if we add Solid.JS on Top of JavaScript, we should need nothing more and nothing extra.

All the features that will make us need Nuxt, Vuetify, Pinia and Vitest---after learning Vue, 

or all the features that will make us need Next, Redux, Styled Components, Jest--after learning React,

should all come as Core in-built features of Solid.JS-- so that once we add Solid.JS on Top of JavaScript, we should need nothing more and nothing extra.

Solid.JS should be the last and final step to learn after JavaScript, and it should take us straight into Production.

Regards.

**Comment by atk** [comment]
Wordpress is a CMS, Python is a programming language. Neither of them are frameworks. "Batteries included" is the quintessential antithesis to what a framework is.

That's not to say it would be impossible to create a full-featured CMS on top of solid (and as far as I know, there are already people working on this tedious task), but that's not what solid is.

**Reply by CordialGit** [reply]
@atk 
The thing is, I have been learning for months, and it's not easy.

Now, Solid is new, and it still has all the opportunity to retrace its steps.

When we develop websites with Django, we almost need nothing more, because of the **Batteries-Included** approach.

Why can't I just learn Solid, and that will be all I ever need?  --in which case all my need for State Management, Server Side Rendering, Routing, Caching, Fetching Data, Consuming API, Bundling, Deployment, and all things needed to have a Production Ready App is all built-in? 

Must I learn Solid, then learn other things on Top of Solid again, before I can get a production ready App with Solid?

Why can't Solid be built to be the last and final bus-stop for me after Vanilla JavaScript?

This is my bone of contention, and this is what I expect Solid to be-- if I'm going to digress from my React and Next.JS journey, to even bother looking at Solid.

**Reply by atk** [reply]
> @atk 
> The thing is, I have been learning for months, and it's not easy.

I have been learning for the last 28 years. It was even harder before the browser wars came to an end and frameworks like solid emerged.

> Now, Solid is new, and it still has all the opportunity to retrace its steps.

Solid is already on its way to 2.0.0. It's been around for years now. That's hardly "new".

> When we develop websites with Django, we almost need nothing more, because of the **Batteries-Included** approach.

If that's what fulfills your requirements completely, I would assume that you need very little frontend interaction, so a framework like Solid will bring you little advantages. If you need batteries-included, choose a CMS, not a framework.

> Why can't I just learn Solid, and that will be all I ever need?  --in which case all my need for State Management, Server Side Rendering, Routing, Caching, Fetching Data, Consuming API, Bundling, Deployment, and all things needed to have a Production Ready App is all built-in? 

Because that's not what a framework is. You're ranting at a cat for not being a horse-drawn carriage.

> Must I learnt Solid, then learn other things on Top of Solid again, before I can get a production ready App with Solid?

That depends on your requirements. That being said, there is Solid-Start, currently in beta, which aims to solve at least a few of your issues, if not all of them.

> Why can't Solid be built to be the last and final bus-stop for me after Vanilla JavaScript?

Because the modern world of development is complex and requirements are manifold and different. You can outsource that complexity to a CMS, at the expense of less performance, accessibility and flexibility if it fulfills your requirements. We won't stop you.

> This is my bone of contention, and this is what I expect Solid to be-- if I'm going to digress from my React and Next journey, to even bother looking at Solid.

Sorry to nitpick, but not Solid is wrong, your expectations are. Solid-Start is going to be more like Next.

**Reply by snnsnn** [reply]
He is leaving similar comments for every tool he uses. It is best to close it.

**Reply by CordialGit** [reply]
@atk 

The fact that you, or the entire JavaScript Community, and all JavaScript Frameworks Community disagree with me, doesn't mean that I'm wrong.

I'm sure that when **John Resig** proposed the idea of **JQuery**, some people told him that he was wasting his time, and JavaScript is _fine as is_.

But when he released his JQuery, that thing took the JavaScript Community like a Tsunami.

Now, what was the **Selling Point** or **Value Proposition** of JQuery? -- it was the Simplicity it brought to writing JavaScript Code. It reduced the amount of lines of JavaScript code that developers needed to write.

Then it solved the **cross-browser JavaScript** problem, where one codebase can be translated across whatever kind of browser there is.

While I've been misunderstood, all I'm simply asking for, is another simplified JavaScript Framework-- which will be "**All-Complete**".

So that, instead of learning Nuxt on Top of Vue, or learning Next on Top of React, or learning SveltKit on Top of Svelt, or learning SolidStart on Top of Solid-- it will just be One Framework that is Complete with all the Features in-built.

In which case, it will come once with everything as Nuxt adds to Vue, or as Next adds to React.

Now, it will be modular in approach. So that if you don't want it's Server Side rendering Features, you don't activate it. On and on like that, will you enable and disable the features you need and don't need.

The fact that many of you have been trained to believe that we must progress from JavaScript to React, then React to Next.   

Or from JavaScript to Vue, then Vue to Nuxt, doesn't mean that this trajectory cannot be broken.

Someone someday(_probably me, when my hand becomes strong in programming_) can come up someday with a trajectory of a direct transition from JavaScript straight into One framework that is all **Turing-and-JavaScript Complete**: which will be the last Framework a developer will need to learn to achieve full Production ready Application, with nothing else needed to be added on top of it.

== One Framework that allows direct transitioning from JavaScript into it, is all **Turing-and-JavaScript Complete**, and is a sufficiently Complete Framework that a developer will learn to achieve full Production ready Web Apps, with nothing else needed to be added on top of it.  ==

This is my reasoning. It is like an expanded concept of the Simplicity Idea of JQuery. Why is this such a bad way to reason?-- that I'm being thrown with trolls and tantrums?

I am simply proposing or suggesting another simplistic way to look at JavaScript Frameworks. I do hope someone someday, will catch this vision, and bring it alive into this world!

**Reply by snnsnn** [reply]
It is not about being agreed or disagreed, it is about time and resources. Problem is half of your demands are mutually exclusive, you have very vague or broad ideas on how those tools work, you put zero effort to achieve those goals other than opening those issues, but the main problem is you are wasting peoples precious time rather than educating yourself. That is why i said we should close this issue because more and more people will try to explain things.

**Reply by CordialGit** [reply]
@snnsnn 

This is all I ask:

== One Framework that allows direct transitioning from JavaScript into it, is all Turing-and-JavaScript Complete, and is a sufficiently Complete Framework that a developer will learn to achieve full Production ready Web Apps, with nothing else needed to be added on top of it. ==

If that's senseless to you, then there's no point. It simply means that I'm with the wrong crowd. No need for further postulations.

**Reply by atk** [reply]
No, you're not wrong because everyone else thinks otherwise. You're wrong because you are confusing a lot of things.

Take jQuery for example. Its value actually was fixing the inconsistencies of DOM interactions especially with older browsers like MSIE6. When MSIE7 was finally buried after the end of the browser wars, the worst offenders were no more, so the only value in jQuery was that people were used to it. It wasn't even a framework, but a toolkit.

A framework is, like the name suggests, only the outline for what can run inside of it. If you have batteries included, it is no longer a framework. That's not necessarily a bad thing. It allows for much more flexibility than for example a CMS and needs not be as opinionated.

You going around asking people to change their framework into a CMS is only a waste of time. If they wanted to create a CMS, they would have done so in the first place. There are options like WordPress, Shopify and Joomla, that should do what you want.

**Reply by atk** [reply]
> This is all I ask:
> == One Framework that allows direct transitioning from JavaScript into it, is all Turing-and-JavaScript Complete, and is a sufficiently Complete Framework that a developer will learn to achieve full Production ready Web Apps, with nothing else needed to be added on top of it. ==

JavaScript already is Turing-complete and you can build production-ready apps in basically any framework available if you take the time to learn them (it helps to know the [underlying concepts](https://dev.to/lexlohr/concepts-behind-modern-frameworks-4m1g)).

> If that's senseless to you, then there's no point. It simply means that I'm with the wrong crowd. No need for further postulations.

It is not only senseless, but unreasonable, because that is not what a framework is. Also, there are no silver bullets in development. Too many people have tried and failed to deliver on that promise. Even if you were a more experienced and prolific developer than all of us combined, you would not be able to create such a solution, because it is a contradiction in itself.

**Reply by samuelstroschein** [reply]
@atk i think you are too technical in your response. whatever jquery, WordPress, etc. are, they help @CordialGit achieve something. namely, build an app/website without 20 different things to learn. imagine the following scenario:

"How do I build a web app?" and someone replies:
 
"You need a server framework - Node | Deno | Bun"
"You need a JS framework - SolidJS | React | Svelte | Vue | Angular"
"You need an app framework - SolidStart | Next | Nuxt | SvelteKit"
"You need auth framework - NextAuth | AuthJS | X"
"You need an i18n framework"
"You need an ..."

Outside of the JS ecosystem the answer is simply "Use Django.", "Use Flutter", "Use Ruby on Rails". I think that's what @CordialGit is referring to and ofc a potential for SolidJS. "How do I build a web app?" -> "Use SolidStart".

**Reply by atk** [reply]
And I think you're not technical enough in your criticism of my respone, @samuelstroschein. There are different sorts of packages for a reason - and that's true outside of the JS ecosystem as well. Saying "Why can't Solid.js should fulfil for me what WordPress does?" is inherently wrong, because it misunderstands the reason for what Solid.js is completely.

It doesn't matter if your CMS is written in PHP, Python or JS. It manages content, because that's what a CMS does. It's not a framework, because it needs to go a lot further than that. Likewise, it doesn't matter if your framework is written in Rust, Ruby on Rails or JS, because at the end of the day, it is just there so that the HTML gets into the DOM at the right place and things are interactive where they should be.

That's why I told @CordialGit that if what he wants fulfils the part of a CMS, he should go search for a CMS instead of criticizing a framework for not being a CMS.

**Reply by samuelstroschein** [reply]
@atk Don't get hung up on the CMS remark. Take a look at my "You need..." list ;) 

I think we can agree that the JS ecosystem is severely fragmented (for reasons). Responding with "but Solid is only a framework and you need X in addition _to build an app_" is exactly what @CordialGit is criticizing. 

Why do we need stuff like Jest/vite/etc? Why do we need all those extra bells and whistles just to build a web app? I know why but explaining to outsiders that "this is how the web ecosystem evolved and Solid is only a framework" will not satisfy potential adopters of SolidJS _that are looking to build an app with one tool only_ :)  

Anyhow, I just wanted to express that @CordialGit message received and I sympathize. The complexity of web apps prevents a "one tool to build apps" solution. But, things are slowly coming together with web standards, node having a test runner now, and more. It will, unfortunately, take time.

**Reply by atk** [reply]
I think you get me wrong again. Also, just about any language ecosystem is similarly fragmented. There is a very good reason to have small separate modules and it boils down to RFC 1925, #5: "It is always possible to aglutenate multiple separate problems into a single complex interdependent solution. In most cases this is a bad idea." - what you gain with just a "one thing does all" solution is reduced options. What you lose is options. What you don't get is a reduced learning curve, because this complex monolithic solution is still comprised of multiple fragments and once you dive deeper, there's still the same amount of things you need to know to make everything work like you want.

What @CordialGit  and you are asking for is a platform that replaces the knowledge of a skilled developer. I have witnessed mutliple attempts to replace developers with code and seen all of them fail. It was true when back-end devs tried to replace front-end devs with Java-to-JavaScript transpilation and it still holds true even in the advent of sophisticated large-language-models. It's not that I as a skilled developer would be against such a platform; I would gladly help developing it. I just realize that it is simply impossible with the current state of tech and realistically will remain so for the next few centuries and thus, it is a fruitless endeavor and a waste of time.

**Reply by Tanay-Verma** [reply]
@CordialGit Just use Angular if you are so hung up on batteries included "frontend 'framework' "

**Comment by CordialGit** [comment]
And by PRs, you mean Public Repositories?

**Reply by Innominus** [reply]
@CordialGit PR meaning Pull Request into the repository.
